core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/DecorrelateUnnest.java

```diff
@@ -0,0 +1,402 @@
+        // If the projection is not necessary for UnnestNode (i.e. it does not produce any unnest symbols), it should be pruned afterwards.
+        PlanNode unnestSource = context.getLookup().resolve(unnestNode.getSource());
+        if (unnestSource instanceof ProjectNode) {
+            ProjectNode sourceProjection = (ProjectNode) unnestSource;
+            input = new ProjectNode(
+                    sourceProjection.getId(),
+                    input,
+                    Assignments.builder()
+                            .putIdentities(input.getOutputSymbols())
+                            .putAll(sourceProjection.getAssignments())
+                            .build());
+        }
+
+        // determine join type for rewritten UnnestNode
+        Type unnestJoinType = LEFT;
+        if (enforceSingleRow.isEmpty() && correlatedJoinNode.getType() == CorrelatedJoinNode.Type.INNER && unnestNode.getJoinType() == INNER) {
+            unnestJoinType = INNER;
+        }
+
+        // make sure that the rewritten node is with ordinality, which might be necessary to restore inner unnest semantics after rewrite.
+        Symbol ordinalitySymbol = unnestNode.getOrdinalitySymbol().orElseGet(() -> context.getSymbolAllocator().newSymbol("ordinality", BIGINT));
+
+        // rewrite correlated join to UnnestNode.
+        UnnestNode rewrittenUnnest = new UnnestNode(
+                context.getIdAllocator().getNextId(),
+                input,
+                input.getOutputSymbols(),
+                unnestNode.getMappings(),
+                Optional.of(ordinalitySymbol),
+                unnestJoinType,
+                Optional.empty());
+
+        // restore all projections, LimitNodes and EnforceSingleRowNode from the subquery
+        PlanNode rewrittenPlan = rewriteNodeSequence(
+                context.getLookup().resolve(correlatedJoinNode.getSubquery()),
+                input.getOutputSymbols(),
+                ordinalitySymbol,
+                uniqueSymbol,
+                rewrittenUnnest,
+                unnestNode.getId(),
+                context.getLookup(),
+                context.getIdAllocator(),
+                context.getSymbolAllocator());
+
+        // restore INNER semantics in case when the original UnnestNode was INNER and the rewritten UnnestNode is LEFT.
+        // In such case, after the rewrite nulls are appended before projections while in the original query they would be appended
+        // after projections, i.e. in EnforceSingleRowNode or in left correlated join. Because after the rewrite nulls are appended
+        // earlier, they might be projected to something else. Project them back to nulls using ordinalitySymbol to distinguish
+        // between unnested rows and synthetic rows added by left unnest.
+        if (unnestNode.getJoinType() == INNER && rewrittenUnnest.getJoinType() == LEFT) {
```

Comment:
please remind reader here that in this case `correlatedJoin.type is == LEFT`

Reply:
It doesn't have to be LEFT. It can also be INNER correlated join + EnforceSingleRowNode. I'll try to clarify it in the comment.